### ISSUE TITLE
Temperature readings wrong around with negative temperatures

### ISSUE THREAD

**Issue by oleo65** [stanvn/zigbee-plant-sensor#9]
I am experiencing a strange behaviour with the reported temperature values in Home Assistant via ZHA integration.

Currently we have negative temperatures outside but the reported temperatures are not. The chart probably best describes the behaviour. 

![Screenshot 2022-12-14 140506](https://user-images.githubusercontent.com/4728549/207822086-2b380312-1398-463c-b883-631f153720f4.png)

At first I suspected a signed/unsigned integer error but the source code seems fine to me. Is there any setting with the SHTC3 sensor which might cause this?

I can rule out a defective device since all three sensor which I placed outside are showing the same behaviour.

**Comment by bbadaboom** [comment]
As I am currently setting up some zigbee plant sensors (thanks for this great project by the way!) I tried to reproduce the situation shown above.
While I can't confirm the weird behavior of the temperature rising, mine also have strage behaviour around the 0°C mark.
When placed outside they move down to just below the 0° mark quite quickly and stay there (moving between -0.3 and -0.7), even though it is actually around -8°C outside.

![image](https://user-images.githubusercontent.com/16547372/208292692-d5f7312f-031c-4ba8-93db-a4f9a70cd580.png)

**Comment by stanvn** [comment]
@oleo65 your are using ZHA I assume?
@bbadaboom what are you using? 

I've place one sensor in the freezer and for me it works fine as you can see in the image (it is not the best freezer). Did you build the image yourself or did you use the prebuild file in the release section?

![Screenshot from 2022-12-18 13-58-31](https://user-images.githubusercontent.com/8735520/208299888-f2438d0a-038e-494d-b9ae-641800729e98.png)

At first I did not expect people to place them outside so I did not optimize the code for large temperature fluctuations. For example the ADC is not re-calibrate when the temperature changes more that 10 degrees. Keep in mind that the clock might also become more inaccurate at lower temperatures, so the battery life might drop when placed outside.

**Comment by oleo65** [comment]
Yes I use the Home Assistant ZHA integration and compiled the firmware myself. 

Your chart shows exactly the same strange behavior around 0 degrees and then dropping further, also below 0. That _is_ exactly the issue I was describing above. The temperature readings are off by a few degrees then, at least way out of the tolerance limits of the SHTC3 sensor.

With a different firmware on the identical hardware I do not experience this issue.

**Comment by stanvn** [comment]
You’re right, I thought it was because I did open the fridge once, but apparently it is the same issue. I’ll investigate it feather, hopefully it is something that can be fixed easily.

**Comment by stanvn** [comment]
After some digging I found a bug in the SHTCx driver from zephyr. I've added a quick fix in the main repo. Could someone verify if this solves the issue?

**Comment by oleo65** [comment]
I might be able to try tomorrow. Thanks for researching this.

**Comment by oleo65** [comment]
I just put a freshly flashed parasite in the freezer and it went down to -20°C. So I would say this is resolved.

**Comment by bbadaboom** [comment]
Haven't managed to compile my own hex file yet so I can't test at the moment.
(Hoping to get there some time around christmas, but so far have no experience with zephyr and I kinda want to know what I am doing, not just blindly installing everything and compiling... ;))

**Comment by stanvn** [comment]
Sorry for the delay due to the holidays, but here is a hex file you can use to test. Let me know if this solves your issue.
https://drive.google.com/file/d/1I_ztPCfgGjZiIEiXSXvrhQuPan2Sl1yL/view?usp=sharing
